### PR TITLE
ast: Extending support for file-level assignments (:=)

### PR DESCRIPTION
**Pull request by johanfylling**
Uppdated support for:
* default values
* rules with `else` keyword
* partial rules
* functions

Fixes: #4555

Signed-off-by: Johan Fylling <johan.dev@fylling.se>

### CODE REVIEW

File: ast/parser.go [hunk 1]
```diff
@@ -727,23 +724,21 @@ func (p *Parser) parseHead(defaultRule bool) *Head {
...
-			return nil
-		}
-
+		s := p.save()
 		p.scan()
 		head.Assign = true
 		head.Value = p.parseTermInfixCall()
 		if head.Value == nil {
-			p.illegal("expected rule value term (e.g., %s := <VALUE> { ... })", head.Name)
+			p.restore(s)
```
Comment (srenatus) [inline ast/parser.go:732]: We restore because we want the failure locations in a specific way, I suppose?

**Review by srenatus — COMMENTED** [review]
Nitpicking aside, it would be nice to have some usage of the new assignments in our formatter test files.

Codestyle wise, I'd have preferred conditionless switch statements over if/else-if/else, but that's really just a personal preference.

File: ast/parser.go (hunk 1 above)
Comment (johanfylling) [reply ast/parser.go:732]: correct, we want to report on the token we just operated on, not the next one.

File: ast/parser.go [hunk 2]
```diff
@@ -639,6 +634,8 @@ func (p *Parser) parseElse(head *Head) *Rule {
 	switch p.s.tok {
 	case tokens.LBrace:
 		rule.Head.Value = BooleanTerm(true)
+	case tokens.Assign:
+		fallthrough
 	case tokens.Unify:
```
[the remark this answers is not in the document]
Comment (johanfylling) [reply ast/parser.go:639]: Didn't know that was an option in go, will fix 👍 .

**Review by srenatus — APPROVED** [review]

**Review by srenatus — COMMENTED** [review]
❓ I guess greping the docs for the previous restriction, and updating them, would be good, too.

**Comment by johanfylling** [comment]
@srenatus , I couldn't find any stated restrictions to update, but I updated the EBNF.

**Comment by srenatus** [comment]
I suppose there're a few usages of `=` that could now be `:=` to yield a consistent picture?

* https://www.openpolicyagent.org/docs/edge/policy-language/#generating-objects
* https://www.openpolicyagent.org/docs/edge/policy-language/#complete-definitions

    `max_memory := 32 { power_users[user] }`, maybe?
* https://www.openpolicyagent.org/docs/edge/policy-language/#functions

I don't know... we don't say anything _wrong_ if we don't update the docs, but if we _want_ people to use `:=` in more places, we should update it. I'm not sure myself, though 🤔

**Comment by johanfylling** [comment]
I was unsure if I should update those places, as we've left rule headers with `=` unification untouched since before. But I don't mind updating all (or most) places I can find that also works with `:=` assignments.

**Review by srenatus — APPROVED** [review]

**Pull request merged**